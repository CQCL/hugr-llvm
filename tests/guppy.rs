use std::{env, path::PathBuf};

use rstest::{fixture, rstest};

struct TestConfig {
    python_bin: PathBuf,
    hugr_llvm_bin: PathBuf,
}

impl TestConfig {
    pub fn new() -> TestConfig {
        let python_bin = env::var("HUGR_LLVM_PYTHON_BIN")
            .map(Into::into)
            .ok()
            .or_else(|| pathsearch::find_executable_in_path("python"))
            .unwrap_or_else(|| panic!("Could not find python in PATH or HUGR_LLVM_PYTHON_BIN"));
        let hugr_llvm_bin = env!("CARGO_BIN_EXE_hugr-llvm").into();
        TestConfig {
            python_bin,
            hugr_llvm_bin,
        }
    }
}

#[fixture]
fn test_config() -> TestConfig {
    TestConfig::new()
}

#[rstest]
fn test_even_odd(test_config: TestConfig) {}
