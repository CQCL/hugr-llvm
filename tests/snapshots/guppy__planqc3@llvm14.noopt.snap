---
source: tests/guppy.rs
info:
  program: hugr-llvm
  args:
    - "--no-opt"
    - "-"
---
success: true
exit_code: 0
----- stdout -----
; ModuleID = 'module'
source_filename = "module"

define i32 @_hl.rus.35(i32 %0, i64 %1) {
alloca_block:
  %"0" = alloca i32, align 4
  %"5_0" = alloca i32, align 4
  %"5_1" = alloca i64, align 8
  %"37_0" = alloca i32, align 4
  %"6_0" = alloca i32, align 4
  %"6_1" = alloca i64, align 8
  %"03" = alloca i32, align 4
  %"7_0" = alloca i64, align 8
  %"7_1" = alloca i64, align 8
  %"7_2" = alloca i32, align 4
  %"10_0" = alloca i64, align 8
  %"10_1" = alloca i64, align 8
  %"10_2" = alloca i32, align 4
  %"11_0" = alloca i32, align 4
  %"12_0" = alloca i64, align 8
  %"12_1" = alloca i64, align 8
  %"12_2" = alloca i32, align 4
  %"12_3" = alloca i32, align 4
  %"15_0" = alloca i64, align 8
  %"15_1" = alloca i64, align 8
  %"15_2" = alloca i32, align 4
  %"15_3" = alloca i32, align 4
  %"16_0" = alloca i32, align 4
  %"17_0" = alloca i64, align 8
  %"17_1" = alloca i64, align 8
  %"17_2" = alloca i32, align 4
  %"40_0" = alloca { {} }, align 8
  %"39_0" = alloca i64, align 8
  %"43_0" = alloca { i32, {}, { i64, i64 } }, align 8
  %"42_0" = alloca { i32, {}, {} }, align 8
  %"021" = alloca { i32, {}, { i64, i64 } }, align 8
  %"45_0" = alloca { i32, {}, { i64, i64 } }, align 8
  %"47_0" = alloca { i32, {}, { i64, i64 } }, align 8
  %"60_1" = alloca { i32, {}, {} }, align 8
  %"58_0" = alloca i32, align 4
  %"51_0" = alloca i32, align 4
  %"52_0" = alloca i32, align 4
  %"49_0" = alloca i32, align 4
  %"50_0" = alloca i32, align 4
  %"53_0" = alloca { { i32, i32 } }, align 8
  %"54_0" = alloca i32, align 4
  %"54_1" = alloca i32, align 4
  %"55_0" = alloca i32, align 4
  %"56_0" = alloca i32, align 4
  %"56_1" = alloca i32, align 4
  %"57_0" = alloca { { i32, i32 } }, align 8
  %"58_1" = alloca i32, align 4
  %"59_0" = alloca i32, align 4
  %"60_0" = alloca i32, align 4
  %"63_0" = alloca { {} }, align 8
  %"73_0" = alloca { i32, { i64, i64 }, {} }, align 8
  %"69_0" = alloca i32, align 4
  %"65_0" = alloca i32, align 4
  %"66_0" = alloca i32, align 4
  %"67_0" = alloca i32, align 4
  %"67_1" = alloca i32, align 4
  %"68_0" = alloca { { i32, i32 } }, align 8
  %"69_1" = alloca i32, align 4
  %"70_0" = alloca i32, align 4
  %"71_0" = alloca i32, align 4
  %"71_1" = alloca { i32, {}, {} }, align 8
  %"083" = alloca { i32, { i64, i64 }, {} }, align 8
  %"75_0" = alloca { i32, { i64, i64 }, {} }, align 8
  %"77_0" = alloca { i32, { i64, i64 }, {} }, align 8
  %"81_0" = alloca { {} }, align 8
  %"80_0" = alloca { {} }, align 8
  %"83_0" = alloca { {} }, align 8
  %"89_0" = alloca { {} }, align 8
  %"88_0" = alloca i64, align 8
  %"85_0" = alloca i32, align 4
  %"87_0" = alloca i64, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store i32 %0, i32* %"5_0", align 4
  store i64 %1, i64* %"5_1", align 4
  %"5_01" = load i32, i32* %"5_0", align 4
  %"5_12" = load i64, i64* %"5_1", align 4
  store i32 %"5_01", i32* %"6_0", align 4
  store i64 %"5_12", i64* %"6_1", align 4
  br label %4

2:                                                ; preds = %4
  %3 = extractvalue { {} } %"40_010", 0
  store i64 %"39_011", i64* %"7_0", align 4
  store i64 %"6_112", i64* %"7_1", align 4
  store i32 %"6_013", i32* %"7_2", align 4
  br label %11

4:                                                ; preds = %entry_block
  %"6_04" = load i32, i32* %"6_0", align 4
  %"6_15" = load i64, i64* %"6_1", align 4
  store { {} } undef, { {} }* %"40_0", align 1
  store i64 0, i64* %"39_0", align 4
  store i32 %"6_04", i32* %"6_0", align 4
  store i64 %"6_15", i64* %"6_1", align 4
  %"40_06" = load { {} }, { {} }* %"40_0", align 1
  %"39_07" = load i64, i64* %"39_0", align 4
  %"6_18" = load i64, i64* %"6_1", align 4
  %"6_09" = load i32, i32* %"6_0", align 4
  store { {} } %"40_06", { {} }* %"40_0", align 1
  store i64 %"39_07", i64* %"39_0", align 4
  store i64 %"6_18", i64* %"6_1", align 4
  store i32 %"6_09", i32* %"6_0", align 4
  %"40_010" = load { {} }, { {} }* %"40_0", align 1
  %"39_011" = load i64, i64* %"39_0", align 4
  %"6_112" = load i64, i64* %"6_1", align 4
  %"6_013" = load i32, i32* %"6_0", align 4
  switch i32 0, label %2 [
  ]

5:                                                ; preds = %cond_exit_43
  %6 = extractvalue { i32, {}, { i64, i64 } } %"43_029", 1
  store i32 %"7_230", i32* %"11_0", align 4
  br label %41

7:                                                ; preds = %cond_exit_43
  %8 = extractvalue { i32, {}, { i64, i64 } } %"43_029", 2
  %9 = extractvalue { i64, i64 } %8, 0
  %10 = extractvalue { i64, i64 } %8, 1
  store i64 %9, i64* %"10_0", align 4
  store i64 %10, i64* %"10_1", align 4
  store i32 %"7_230", i32* %"10_2", align 4
  br label %21

11:                                               ; preds = %68, %62, %2
  %"7_015" = load i64, i64* %"7_0", align 4
  %"7_116" = load i64, i64* %"7_1", align 4
  %"7_217" = load i32, i32* %"7_2", align 4
  store i64 %"7_015", i64* %"7_0", align 4
  store i64 %"7_116", i64* %"7_1", align 4
  store i32 %"7_217", i32* %"7_2", align 4
  %"7_018" = load i64, i64* %"7_0", align 4
  %"7_119" = load i64, i64* %"7_1", align 4
  %12 = icmp slt i64 %"7_018", %"7_119"
  %13 = select i1 %12, { i32, {}, {} } { i32 1, {} poison, {} undef }, { i32, {}, {} } { i32 0, {} undef, {} poison }
  store { i32, {}, {} } %13, { i32, {}, {} }* %"42_0", align 4
  %"42_020" = load { i32, {}, {} }, { i32, {}, {} }* %"42_0", align 4
  %14 = extractvalue { i32, {}, {} } %"42_020", 0
  %15 = extractvalue { i32, {}, {} } %"42_020", 1
  %16 = extractvalue { i32, {}, {} } %"42_020", 2
  switch i32 %14, label %cond_43_case_0 [
    i32 1, label %cond_43_case_1
  ]

17:                                               ; preds = %21
  %18 = extractvalue { i32, {}, {} } %"60_155", 1
  store i64 %"10_056", i64* %"15_0", align 4
  store i64 %"10_157", i64* %"15_1", align 4
  store i32 %"58_058", i32* %"15_2", align 4
  store i32 %"10_259", i32* %"15_3", align 4
  br label %64

19:                                               ; preds = %21
  %20 = extractvalue { i32, {}, {} } %"60_155", 2
  store i64 %"10_056", i64* %"12_0", align 4
  store i64 %"10_157", i64* %"12_1", align 4
  store i32 %"58_058", i32* %"12_2", align 4
  store i32 %"10_259", i32* %"12_3", align 4
  br label %48

21:                                               ; preds = %7
  %"10_031" = load i64, i64* %"10_0", align 4
  %"10_132" = load i64, i64* %"10_1", align 4
  %"10_233" = load i32, i32* %"10_2", align 4
  %qalloc_call = call i32 @___qalloc()
  store i32 %qalloc_call, i32* %"51_0", align 4
  %"51_034" = load i32, i32* %"51_0", align 4
  %h_call = call i32 @___h(i32 %"51_034")
  store i32 %h_call, i32* %"52_0", align 4
  %qalloc_call35 = call i32 @___qalloc.1()
  store i32 %qalloc_call35, i32* %"49_0", align 4
  %"49_036" = load i32, i32* %"49_0", align 4
  %h_call37 = call i32 @___h.2(i32 %"49_036")
  store i32 %h_call37, i32* %"50_0", align 4
  %"50_038" = load i32, i32* %"50_0", align 4
  %"52_039" = load i32, i32* %"52_0", align 4
  %22 = insertvalue { i32, i32 } undef, i32 %"50_038", 0
  %23 = insertvalue { i32, i32 } %22, i32 %"52_039", 1
  %24 = insertvalue { { i32, i32 } } poison, { i32, i32 } %23, 0
  store { { i32, i32 } } %24, { { i32, i32 } }* %"53_0", align 4
  %"53_040" = load { { i32, i32 } }, { { i32, i32 } }* %"53_0", align 4
  %25 = extractvalue { { i32, i32 } } %"53_040", 0
  %26 = extractvalue { i32, i32 } %25, 0
  %27 = extractvalue { i32, i32 } %25, 1
  store i32 %26, i32* %"54_0", align 4
  store i32 %27, i32* %"54_1", align 4
  %"54_041" = load i32, i32* %"54_0", align 4
  %tdg_call = call i32 @___tdg(i32 %"54_041")
  store i32 %tdg_call, i32* %"55_0", align 4
  %"54_142" = load i32, i32* %"54_1", align 4
  %"55_043" = load i32, i32* %"55_0", align 4
  %rz_call = call { i32, i32 } @___cx(i32 %"54_142", i32 %"55_043")
  %28 = extractvalue { i32, i32 } %rz_call, 0
  %29 = extractvalue { i32, i32 } %rz_call, 1
  store i32 %28, i32* %"56_0", align 4
  store i32 %29, i32* %"56_1", align 4
  %"56_044" = load i32, i32* %"56_0", align 4
  %"56_145" = load i32, i32* %"56_1", align 4
  %30 = insertvalue { i32, i32 } undef, i32 %"56_044", 0
  %31 = insertvalue { i32, i32 } %30, i32 %"56_145", 1
  %32 = insertvalue { { i32, i32 } } poison, { i32, i32 } %31, 0
  store { { i32, i32 } } %32, { { i32, i32 } }* %"57_0", align 4
  %"57_046" = load { { i32, i32 } }, { { i32, i32 } }* %"57_0", align 4
  %33 = extractvalue { { i32, i32 } } %"57_046", 0
  %34 = extractvalue { i32, i32 } %33, 0
  %35 = extractvalue { i32, i32 } %33, 1
  store i32 %34, i32* %"58_0", align 4
  store i32 %35, i32* %"58_1", align 4
  %"58_147" = load i32, i32* %"58_1", align 4
  %t_call = call i32 @___t(i32 %"58_147")
  store i32 %t_call, i32* %"59_0", align 4
  %"59_048" = load i32, i32* %"59_0", align 4
  %measure_call = call { i32, { i32, {}, {} } } @___measure(i32 %"59_048")
  %36 = extractvalue { i32, { i32, {}, {} } } %measure_call, 0
  %37 = extractvalue { i32, { i32, {}, {} } } %measure_call, 1
  store i32 %36, i32* %"60_0", align 4
  store { i32, {}, {} } %37, { i32, {}, {} }* %"60_1", align 4
  %"60_049" = load i32, i32* %"60_0", align 4
  call void @___qfree(i32 %"60_049")
  store i64 %"10_031", i64* %"10_0", align 4
  store i64 %"10_132", i64* %"10_1", align 4
  store i32 %"10_233", i32* %"10_2", align 4
  %"60_150" = load { i32, {}, {} }, { i32, {}, {} }* %"60_1", align 4
  %"10_051" = load i64, i64* %"10_0", align 4
  %"10_152" = load i64, i64* %"10_1", align 4
  %"58_053" = load i32, i32* %"58_0", align 4
  %"10_254" = load i32, i32* %"10_2", align 4
  store { i32, {}, {} } %"60_150", { i32, {}, {} }* %"60_1", align 4
  store i64 %"10_051", i64* %"10_0", align 4
  store i64 %"10_152", i64* %"10_1", align 4
  store i32 %"58_053", i32* %"58_0", align 4
  store i32 %"10_254", i32* %"10_2", align 4
  %"60_155" = load { i32, {}, {} }, { i32, {}, {} }* %"60_1", align 4
  %"10_056" = load i64, i64* %"10_0", align 4
  %"10_157" = load i64, i64* %"10_1", align 4
  %"58_058" = load i32, i32* %"58_0", align 4
  %"10_259" = load i32, i32* %"10_2", align 4
  %38 = extractvalue { i32, {}, {} } %"60_155", 0
  switch i32 %38, label %17 [
    i32 1, label %19
  ]

39:                                               ; preds = %41
  %40 = extractvalue { {} } %"63_063", 0
  store i32 %"11_064", i32* %"03", align 4
  br label %72

41:                                               ; preds = %65, %5
  %"11_060" = load i32, i32* %"11_0", align 4
  store { {} } undef, { {} }* %"63_0", align 1
  store i32 %"11_060", i32* %"11_0", align 4
  %"63_061" = load { {} }, { {} }* %"63_0", align 1
  %"11_062" = load i32, i32* %"11_0", align 4
  store { {} } %"63_061", { {} }* %"63_0", align 1
  store i32 %"11_062", i32* %"11_0", align 4
  %"63_063" = load { {} }, { {} }* %"63_0", align 1
  %"11_064" = load i32, i32* %"11_0", align 4
  switch i32 0, label %39 [
  ]

42:                                               ; preds = %cond_exit_73
  %43 = extractvalue { i32, { i64, i64 }, {} } %"73_092", 1
  %44 = extractvalue { i64, i64 } %43, 0
  %45 = extractvalue { i64, i64 } %43, 1
  store i64 %44, i64* %"17_0", align 4
  store i64 %45, i64* %"17_1", align 4
  store i32 %"69_093", i32* %"17_2", align 4
  br label %70

46:                                               ; preds = %cond_exit_73
  %47 = extractvalue { i32, { i64, i64 }, {} } %"73_092", 2
  store i32 %"69_093", i32* %"16_0", align 4
  br label %67

48:                                               ; preds = %19
  %"12_065" = load i64, i64* %"12_0", align 4
  %"12_166" = load i64, i64* %"12_1", align 4
  %"12_267" = load i32, i32* %"12_2", align 4
  %"12_368" = load i32, i32* %"12_3", align 4
  store i64 %"12_065", i64* %"12_0", align 4
  store i64 %"12_166", i64* %"12_1", align 4
  store i32 %"12_267", i32* %"12_2", align 4
  store i32 %"12_368", i32* %"12_3", align 4
  %"12_369" = load i32, i32* %"12_3", align 4
  %t_call70 = call i32 @___t.3(i32 %"12_369")
  store i32 %t_call70, i32* %"65_0", align 4
  %"65_071" = load i32, i32* %"65_0", align 4
  %z_call = call i32 @___z(i32 %"65_071")
  store i32 %z_call, i32* %"66_0", align 4
  %"66_072" = load i32, i32* %"66_0", align 4
  %"12_273" = load i32, i32* %"12_2", align 4
  %rz_call74 = call { i32, i32 } @___cx.4(i32 %"66_072", i32 %"12_273")
  %49 = extractvalue { i32, i32 } %rz_call74, 0
  %50 = extractvalue { i32, i32 } %rz_call74, 1
  store i32 %49, i32* %"67_0", align 4
  store i32 %50, i32* %"67_1", align 4
  %"67_075" = load i32, i32* %"67_0", align 4
  %"67_176" = load i32, i32* %"67_1", align 4
  %51 = insertvalue { i32, i32 } undef, i32 %"67_075", 0
  %52 = insertvalue { i32, i32 } %51, i32 %"67_176", 1
  %53 = insertvalue { { i32, i32 } } poison, { i32, i32 } %52, 0
  store { { i32, i32 } } %53, { { i32, i32 } }* %"68_0", align 4
  %"68_077" = load { { i32, i32 } }, { { i32, i32 } }* %"68_0", align 4
  %54 = extractvalue { { i32, i32 } } %"68_077", 0
  %55 = extractvalue { i32, i32 } %54, 0
  %56 = extractvalue { i32, i32 } %54, 1
  store i32 %55, i32* %"69_0", align 4
  store i32 %56, i32* %"69_1", align 4
  %"69_178" = load i32, i32* %"69_1", align 4
  %t_call79 = call i32 @___t.5(i32 %"69_178")
  store i32 %t_call79, i32* %"70_0", align 4
  %"70_080" = load i32, i32* %"70_0", align 4
  %measure_call81 = call { i32, { i32, {}, {} } } @___measure.6(i32 %"70_080")
  %57 = extractvalue { i32, { i32, {}, {} } } %measure_call81, 0
  %58 = extractvalue { i32, { i32, {}, {} } } %measure_call81, 1
  store i32 %57, i32* %"71_0", align 4
  store { i32, {}, {} } %58, { i32, {}, {} }* %"71_1", align 4
  %"71_182" = load { i32, {}, {} }, { i32, {}, {} }* %"71_1", align 4
  %59 = extractvalue { i32, {}, {} } %"71_182", 0
  %60 = extractvalue { i32, {}, {} } %"71_182", 1
  %61 = extractvalue { i32, {}, {} } %"71_182", 2
  switch i32 %59, label %cond_73_case_0 [
    i32 1, label %cond_73_case_1
  ]

62:                                               ; preds = %64
  %63 = extractvalue { {} } %"81_0103", 0
  store i64 %"15_0104", i64* %"7_0", align 4
  store i64 %"15_1105", i64* %"7_1", align 4
  store i32 %"15_3106", i32* %"7_2", align 4
  br label %11

64:                                               ; preds = %17
  %"15_094" = load i64, i64* %"15_0", align 4
  %"15_195" = load i64, i64* %"15_1", align 4
  %"15_296" = load i32, i32* %"15_2", align 4
  %"15_397" = load i32, i32* %"15_3", align 4
  store { {} } undef, { {} }* %"81_0", align 1
  store { {} } undef, { {} }* %"80_0", align 1
  store i64 %"15_094", i64* %"15_0", align 4
  store i64 %"15_195", i64* %"15_1", align 4
  store i32 %"15_296", i32* %"15_2", align 4
  store i32 %"15_397", i32* %"15_3", align 4
  %"81_098" = load { {} }, { {} }* %"81_0", align 1
  %"15_099" = load i64, i64* %"15_0", align 4
  %"15_1100" = load i64, i64* %"15_1", align 4
  %"15_3101" = load i32, i32* %"15_3", align 4
  store { {} } %"81_098", { {} }* %"81_0", align 1
  store i64 %"15_099", i64* %"15_0", align 4
  store i64 %"15_1100", i64* %"15_1", align 4
  store i32 %"15_3101", i32* %"15_3", align 4
  %"15_2102" = load i32, i32* %"15_2", align 4
  call void @___qfree.8(i32 %"15_2102")
  %"81_0103" = load { {} }, { {} }* %"81_0", align 1
  %"15_0104" = load i64, i64* %"15_0", align 4
  %"15_1105" = load i64, i64* %"15_1", align 4
  %"15_3106" = load i32, i32* %"15_3", align 4
  switch i32 0, label %62 [
  ]

65:                                               ; preds = %67
  %66 = extractvalue { {} } %"83_0110", 0
  store i32 %"16_0111", i32* %"11_0", align 4
  br label %41

67:                                               ; preds = %46
  %"16_0107" = load i32, i32* %"16_0", align 4
  store { {} } undef, { {} }* %"83_0", align 1
  store i32 %"16_0107", i32* %"16_0", align 4
  %"83_0108" = load { {} }, { {} }* %"83_0", align 1
  %"16_0109" = load i32, i32* %"16_0", align 4
  store { {} } %"83_0108", { {} }* %"83_0", align 1
  store i32 %"16_0109", i32* %"16_0", align 4
  %"83_0110" = load { {} }, { {} }* %"83_0", align 1
  %"16_0111" = load i32, i32* %"16_0", align 4
  switch i32 0, label %65 [
  ]

68:                                               ; preds = %70
  %69 = extractvalue { {} } %"89_0122", 0
  store i64 %"88_0123", i64* %"7_0", align 4
  store i64 %"17_1124", i64* %"7_1", align 4
  store i32 %"85_0125", i32* %"7_2", align 4
  br label %11

70:                                               ; preds = %42
  %"17_0112" = load i64, i64* %"17_0", align 4
  %"17_1113" = load i64, i64* %"17_1", align 4
  %"17_2114" = load i32, i32* %"17_2", align 4
  store { {} } undef, { {} }* %"89_0", align 1
  store i64 1, i64* %"87_0", align 4
  store i64 %"17_0112", i64* %"17_0", align 4
  store i64 %"17_1113", i64* %"17_1", align 4
  store i32 %"17_2114", i32* %"17_2", align 4
  %"17_2115" = load i32, i32* %"17_2", align 4
  %x_call = call i32 @___x(i32 %"17_2115")
  store i32 %x_call, i32* %"85_0", align 4
  %"17_0116" = load i64, i64* %"17_0", align 4
  %"87_0117" = load i64, i64* %"87_0", align 4
  %71 = add i64 %"17_0116", %"87_0117"
  store i64 %71, i64* %"88_0", align 4
  %"89_0118" = load { {} }, { {} }* %"89_0", align 1
  %"88_0119" = load i64, i64* %"88_0", align 4
  %"17_1120" = load i64, i64* %"17_1", align 4
  %"85_0121" = load i32, i32* %"85_0", align 4
  store { {} } %"89_0118", { {} }* %"89_0", align 1
  store i64 %"88_0119", i64* %"88_0", align 4
  store i64 %"17_1120", i64* %"17_1", align 4
  store i32 %"85_0121", i32* %"85_0", align 4
  %"89_0122" = load { {} }, { {} }* %"89_0", align 1
  %"88_0123" = load i64, i64* %"88_0", align 4
  %"17_1124" = load i64, i64* %"17_1", align 4
  %"85_0125" = load i32, i32* %"85_0", align 4
  switch i32 0, label %68 [
  ]

72:                                               ; preds = %39
  %"014" = load i32, i32* %"03", align 4
  store i32 %"014", i32* %"37_0", align 4
  %"37_0126" = load i32, i32* %"37_0", align 4
  store i32 %"37_0126", i32* %"0", align 4
  %"0127" = load i32, i32* %"0", align 4
  ret i32 %"0127"

cond_43_case_0:                                   ; preds = %11
  store { i32, {}, { i64, i64 } } { i32 0, {} undef, { i64, i64 } poison }, { i32, {}, { i64, i64 } }* %"45_0", align 4
  %"45_023" = load { i32, {}, { i64, i64 } }, { i32, {}, { i64, i64 } }* %"45_0", align 4
  store { i32, {}, { i64, i64 } } %"45_023", { i32, {}, { i64, i64 } }* %"021", align 4
  br label %cond_exit_43

cond_43_case_1:                                   ; preds = %11
  %"7_024" = load i64, i64* %"7_0", align 4
  %"7_125" = load i64, i64* %"7_1", align 4
  %73 = insertvalue { i64, i64 } undef, i64 %"7_024", 0
  %74 = insertvalue { i64, i64 } %73, i64 %"7_125", 1
  %75 = insertvalue { i32, {}, { i64, i64 } } { i32 1, {} poison, { i64, i64 } poison }, { i64, i64 } %74, 2
  store { i32, {}, { i64, i64 } } %75, { i32, {}, { i64, i64 } }* %"47_0", align 4
  %"47_026" = load { i32, {}, { i64, i64 } }, { i32, {}, { i64, i64 } }* %"47_0", align 4
  store { i32, {}, { i64, i64 } } %"47_026", { i32, {}, { i64, i64 } }* %"021", align 4
  br label %cond_exit_43

cond_exit_43:                                     ; preds = %cond_43_case_1, %cond_43_case_0
  %"022" = load { i32, {}, { i64, i64 } }, { i32, {}, { i64, i64 } }* %"021", align 4
  store { i32, {}, { i64, i64 } } %"022", { i32, {}, { i64, i64 } }* %"43_0", align 4
  %"43_027" = load { i32, {}, { i64, i64 } }, { i32, {}, { i64, i64 } }* %"43_0", align 4
  %"7_228" = load i32, i32* %"7_2", align 4
  store { i32, {}, { i64, i64 } } %"43_027", { i32, {}, { i64, i64 } }* %"43_0", align 4
  store i32 %"7_228", i32* %"7_2", align 4
  %"43_029" = load { i32, {}, { i64, i64 } }, { i32, {}, { i64, i64 } }* %"43_0", align 4
  %"7_230" = load i32, i32* %"7_2", align 4
  %76 = extractvalue { i32, {}, { i64, i64 } } %"43_029", 0
  switch i32 %76, label %5 [
    i32 1, label %7
  ]

cond_73_case_0:                                   ; preds = %48
  %"12_085" = load i64, i64* %"12_0", align 4
  %"12_186" = load i64, i64* %"12_1", align 4
  %77 = insertvalue { i64, i64 } undef, i64 %"12_085", 0
  %78 = insertvalue { i64, i64 } %77, i64 %"12_186", 1
  %79 = insertvalue { i32, { i64, i64 }, {} } { i32 0, { i64, i64 } poison, {} poison }, { i64, i64 } %78, 1
  store { i32, { i64, i64 }, {} } %79, { i32, { i64, i64 }, {} }* %"75_0", align 4
  %"75_087" = load { i32, { i64, i64 }, {} }, { i32, { i64, i64 }, {} }* %"75_0", align 4
  store { i32, { i64, i64 }, {} } %"75_087", { i32, { i64, i64 }, {} }* %"083", align 4
  br label %cond_exit_73

cond_73_case_1:                                   ; preds = %48
  store { i32, { i64, i64 }, {} } { i32 1, { i64, i64 } poison, {} undef }, { i32, { i64, i64 }, {} }* %"77_0", align 4
  %"77_088" = load { i32, { i64, i64 }, {} }, { i32, { i64, i64 }, {} }* %"77_0", align 4
  store { i32, { i64, i64 }, {} } %"77_088", { i32, { i64, i64 }, {} }* %"083", align 4
  br label %cond_exit_73

cond_exit_73:                                     ; preds = %cond_73_case_1, %cond_73_case_0
  %"084" = load { i32, { i64, i64 }, {} }, { i32, { i64, i64 }, {} }* %"083", align 4
  store { i32, { i64, i64 }, {} } %"084", { i32, { i64, i64 }, {} }* %"73_0", align 4
  %"73_089" = load { i32, { i64, i64 }, {} }, { i32, { i64, i64 }, {} }* %"73_0", align 4
  %"69_090" = load i32, i32* %"69_0", align 4
  store { i32, { i64, i64 }, {} } %"73_089", { i32, { i64, i64 }, {} }* %"73_0", align 4
  store i32 %"69_090", i32* %"69_0", align 4
  %"71_091" = load i32, i32* %"71_0", align 4
  call void @___qfree.7(i32 %"71_091")
  %"73_092" = load { i32, { i64, i64 }, {} }, { i32, { i64, i64 }, {} }* %"73_0", align 4
  %"69_093" = load i32, i32* %"69_0", align 4
  %80 = extractvalue { i32, { i64, i64 }, {} } %"73_092", 0
  switch i32 %80, label %42 [
    i32 1, label %46
  ]
}

declare i32 @___qalloc()

declare i32 @___h(i32)

declare i32 @___qalloc.1()

declare i32 @___h.2(i32)

declare i32 @___tdg(i32)

declare { i32, i32 } @___cx(i32, i32)

declare i32 @___t(i32)

declare { i32, { i32, {}, {} } } @___measure(i32)

declare void @___qfree(i32)

declare i32 @___t.3(i32)

declare i32 @___z(i32)

declare { i32, i32 } @___cx.4(i32, i32)

declare i32 @___t.5(i32)

declare { i32, { i32, {}, {} } } @___measure.6(i32)

declare void @___qfree.7(i32)

declare void @___qfree.8(i32)

declare i32 @___x(i32)

define { i32, {}, {} } @_hl.main.36() {
alloca_block:
  %"0" = alloca { i32, {}, {} }, align 8
  %"90_0" = alloca { i32, {}, {} }, align 8
  %"01" = alloca { i32, {}, {} }, align 8
  %"97_0" = alloca { {} }, align 8
  %"95_1" = alloca { i32, {}, {} }, align 8
  %"93_0" = alloca i64, align 8
  %"91_0" = alloca i32, align 4
  %"94_0" = alloca i32, align 4
  %"95_0" = alloca i32, align 4
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  br label %2

0:                                                ; preds = %2
  %1 = extractvalue { {} } %"97_08", 0
  store { i32, {}, {} } %"95_19", { i32, {}, {} }* %"01", align 4
  br label %6

2:                                                ; preds = %entry_block
  store { {} } undef, { {} }* %"97_0", align 1
  store i64 100, i64* %"93_0", align 4
  %qalloc_call = call i32 @___qalloc.9()
  store i32 %qalloc_call, i32* %"91_0", align 4
  %"91_02" = load i32, i32* %"91_0", align 4
  %"93_03" = load i64, i64* %"93_0", align 4
  %3 = call i32 @_hl.rus.35(i32 %"91_02", i64 %"93_03")
  store i32 %3, i32* %"94_0", align 4
  %"94_04" = load i32, i32* %"94_0", align 4
  %measure_call = call { i32, { i32, {}, {} } } @___measure.10(i32 %"94_04")
  %4 = extractvalue { i32, { i32, {}, {} } } %measure_call, 0
  %5 = extractvalue { i32, { i32, {}, {} } } %measure_call, 1
  store i32 %4, i32* %"95_0", align 4
  store { i32, {}, {} } %5, { i32, {}, {} }* %"95_1", align 4
  %"97_05" = load { {} }, { {} }* %"97_0", align 1
  %"95_16" = load { i32, {}, {} }, { i32, {}, {} }* %"95_1", align 4
  store { {} } %"97_05", { {} }* %"97_0", align 1
  store { i32, {}, {} } %"95_16", { i32, {}, {} }* %"95_1", align 4
  %"95_07" = load i32, i32* %"95_0", align 4
  call void @___qfree.11(i32 %"95_07")
  %"97_08" = load { {} }, { {} }* %"97_0", align 1
  %"95_19" = load { i32, {}, {} }, { i32, {}, {} }* %"95_1", align 4
  switch i32 0, label %0 [
  ]

6:                                                ; preds = %0
  %"010" = load { i32, {}, {} }, { i32, {}, {} }* %"01", align 4
  store { i32, {}, {} } %"010", { i32, {}, {} }* %"90_0", align 4
  %"90_011" = load { i32, {}, {} }, { i32, {}, {} }* %"90_0", align 4
  store { i32, {}, {} } %"90_011", { i32, {}, {} }* %"0", align 4
  %"012" = load { i32, {}, {} }, { i32, {}, {} }* %"0", align 4
  ret { i32, {}, {} } %"012"
}

declare i32 @___qalloc.9()

declare { i32, { i32, {}, {} } } @___measure.10(i32)

declare void @___qfree.11(i32)


----- stderr -----
HUGR valid!
